### PR TITLE
HCIbox - 23h2 support

### LINKED ISSUE #2304: HCIBox: 23H2 not supported
I have deployed the Jumpstart HCI box. The deployment itself was successful, and the bridge and custom locations were created, but the 2 VM images, as well as any additional marketplace image I try to add, show the error below. I've seen previous comments here that the HCIBox experience is being updated to support 23H2. Is that the cause of this error, and is there any timeline for the Jumpstart HCI box to work again? @dkirby-ms 

{
  "error": {
    "code": "NotSupported",
    "message": "Supported OS Versions : 23H2 or higher. Node(s) not supported : AzSHOST1,AzSHOST2. : Not Supported"
  }
}

image

### PR DESCRIPTION
**Pull request by dkirby-ms**
Adds support for Azure Stack HCI 23H2

* 23h2 OS for HCI nodes
* Cloud deployment replaces older Azure PowerShell-based deployment
* Cloud deployment includes 23h2 Resource bridge and AKS management
* Codebase refactoring and optimizations

Closes #2304

### CODE REVIEW

File: azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1 [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+$storageBIPs = "[ "
+$count = 0
+foreach ($node in $HCIBoxConfig.NodeHostConfig) {
+    if ($count -gt 0) {
+        $physicalNodesSettings += ", "
+        $storageAIPs += ", "
+        $storageBIPs += ", "
+    }
+    $physicalNodesSettings += "{ ""name"": ""$($node.Hostname)"", ""ipv4Address"": ""$($node.IP.Split("/")[0])"" }"
+   # $storageAIPs += "{ ""PhysicalNode"": ""$($node.Hostname)"", ""IPv4Address"": ""$($node.StorageAIP)"", ""SubnetMask"": ""$($HCIBoxConfig.storageAsubnet)"" }"
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1:85]: Can this be removed if not needed?

File: azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1 [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+(Get-Content -Path $hciParams) -replace 'startingIp-staging', $HCIBoxConfig.clusterIpRangeStart | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'endingIp-staging', $HCIBoxConfig.clusterIpRangeEnd | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'dnsServers-staging', $dns | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'keyVaultName-staging', $keyVaultName | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'secretsLocation-staging', $secretsLocation | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'physicalNodesSettings-staging', $physicalNodesSettings | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'ClusterWitnessStorageAccountName-staging', $env:stagingStorageAccountName | Set-Content -Path $hciParams
+(Get-Content -Path $hciParams) -replace 'diagnosticStorageAccountName-staging', $diagnosticsStorageName | Set-Content -Path $hciParams
+#(Get-Content -Path $hciParams) -replace 'storageNetworkA-staging', $storageAIPs | Set-Content -Path $hciParams
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1:117]: Can this be removed if not needed?

File: azure_jumpstart_hcibox/artifacts/hci.json
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/hci.json:1]: Is it an option to decompile this template into Bicep (for better readability)? Or does the deployment command expect/support native ARM templates?

Example:
```bicep
@description('First must pass Validate prior running Deploy')
@allowed([
  'Validate'
  'Deploy'
])
param deploymentMode string = 'Validate'

@description('The KeyVault name used to store the secrets.')
param keyVaultName string = 'hcibox-kv-[substring(newGuid(),0,4)]'
param softDeleteRetentionDays int = 30

@description('The name of the storage account used for KV audit logs')
param diagnosticStorageAccountName string

@description('Specifies the number of days that logs are gonna be kept. If you do not want to apply any retention policy and retain data forever, set value to 0.')
@minValue(0)
@maxValue(365)
param logsRetentionInDays int = 30

@description('Storage Account type')
@allowed([
  'Premium_LRS'
  'Premium_ZRS'
  'Standard_GRS'
  'Standard_GZRS'
  'Standard_LRS'
  'Standard_RAGRS'
  'Standard_RAGZRS'
  'Standard_ZRS'
])
param storageAccountType string = 'Standard_LRS'

@description('The fqdn of your KeyVault')
param secretsLocation string = ''

@description('The existing storage account name used for the cluster witness')
param ClusterWitnessStorageAccountName string = ''

@minLength(3)
@maxLength(24)
param clusterName string
param location string = resourceGroup().location
param tenantId string = subscription().tenantId

@description('The name can not be changed')
@allowed([
  'LocalAdminCredential'
])
param localAdminSecretName string = 'LocalAdminCredential'

@secure()
param localAdminSecretValue string

@description('The name can not be changed')
@allowed([
  'AzureStackLCMUserCredential'
])
param domainAdminSecretName string = 'AzureStackLCMUserCredential'

@secure()
param domainAdminSecretValue string

@description('The name can not be changed')
@allowed([
  'DefaultARBApplication'
])
param arbDeploymentSpnName string = 'DefaultARBApplication'

@secure()
param arbDeploymentSpnValue string

@description('The name can not be changed')
@allowed([
  'WitnessStorageKey'
])
param storageWitnessName string = 'WitnessStorageKey'

@secure()
param storageWitnessValue string

@description('The api version for deploying a hci cluster')
param apiVersion string = ''

@description('The arc for server node Ids of the hci cluster')
param arcNodeResourceIds array = []

@description('The domain name of the Active Directory Domain Services')
param domainFqdn string = ''

@description('The ADFS name prefix')
param namingPrefix string = ''

@description('The ADDS OU path')
param adouPath string = ''

@description('The security level data for deploying a hci cluster')
@allowed([
  'Recommended'
  'Customized'
])
param securityLevel string = 'Recommended'

@description('The security setting driftControlEnforced data for deploying a hci cluster')
@allowed([
  true
  false
])
param driftControlEnforced bool = true

@description('The security setting credentialGuardEnforced data for deploying a hci cluster')
@allowed([
  true
  false
])
param credentialGuardEnforced bool = true

@description('The security setting smbSigningEnforced data for deploying a hci cluster')
@allowed([
  true
  false
])
param smbSigningEnforced bool = true

@description('The security setting smbClusterEncryption data for deploying a hci cluster')
@allowed([
  true
  false
])
param smbClusterEncryption bool = false

@description('The security setting bitlockerBootVolume data for deploying a hci cluster')
@allowed([
  true
  false
])
param bitlockerBootVolume bool = true

@description('The security setting bitlockerDataVolumes data for deploying a hci cluster')
@allowed([
  true
  false
])
param bitlockerDataVolumes bool = true

@description('The security setting wdacEnforced data for deploying a hci cluster')
@allowed([
  true
  false
])
param wdacEnforced bool = true

@description('The metrics data for deploying a hci cluster')
@allowed([
  true
  false
])
param streamingDataClient bool = true

@description('The location data for deploying a hci cluster')
@allowed([
  true
  false
])
param euLocation bool = false

@description('The diagnostic data for deploying a hci cluster')
@allowed([
  true
  false
])
param episodicDataUpload bool = true

@description('The storage volume configuration mode')
@allowed([
  'Express'
  'InfraOnly'
  'KeepStorage'
])
param configurationMode string = 'Express'

@description('The subnet mask for deploying a hci cluster')
param subnetMask string = ''

@description('The default gateway for deploying a hci cluster')
param defaultGateway string = ''

@description('The starting ip address for deploying a hci cluster')
param startingIPAddress string = ''

@description('The ending ip address for deploying a hci cluster')
param endingIPAddress string = ''

@description('The dns servers for deploying a hci cluster')
param dnsServers array = [
  ''
]

@description('The physical nodes settings for deploying a hci cluster')
param physicalNodesSettings array

@description('The networking type for deploying a hci cluster')
@allowed([
  'switchedMultiServerDeployment'
  'switchlessMultiServerDeployment'
  'singleServerDeployment'
])
param networkingType string = ''

@description('The intent list for deploying a hci cluster')
param intentList array = []

@description('The storage network list for deploying a hci cluster')
param storageNetworkList array = []

@description('The storage connectivity switchless value for deploying a hci cluster')
param storageConnectivitySwitchless bool = false

@description('The custom location for deploying a hci cluster')
param customLocation string = ''

resource diagnosticStorageAccount 'Microsoft.Storage/storageAccounts@2021-01-01' = {
  name: diagnosticStorageAccountName
  location: location
  sku: {
    name: storageAccountType
    tier: 'Standard'
  }
  kind: 'StorageV2'
  properties: {
    supportsHttpsTrafficOnly: true
  }
}

resource keyVault 'Microsoft.KeyVault/vaults@2021-06-01-preview' = {
  name: keyVaultName
  location: location
  properties: {
    enabledForDeployment: true
    enabledForTemplateDeployment: true
    enabledForDiskEncryption: true
    enableSoftDelete: true
    softDeleteRetentionInDays: softDeleteRetentionDays
    enableRbacAuthorization: true
    publicNetworkAccess: 'Enabled'
    accessPolicies: []
    tenantId: tenantId
    sku: {
      name: 'standard'
      family: 'A'
    }
  }
  dependsOn: [
    diagnosticStorageAccount
  ]
}

resource keyVaultName_Microsoft_Insights_service 'Microsoft.KeyVault/vaults/providers/diagnosticsettings@2016-09-01' = {
  name: '${keyVaultName}/Microsoft.Insights/service'
  location: resourceGroup().location
  properties: {
    storageAccountId: diagnosticStorageAccount.id
    logs: [
      {
        category: 'AuditEvent'
        enabled: true
        retentionPolicy: {
          enabled: true
          days: logsRetentionInDays
        }
      }
    ]
  }
  dependsOn: [
    keyVault

  ]
}

resource cluster 'Microsoft.AzureStackHCI/clusters@2023-08-01-preview' = if (deploymentMode == 'Validate') {
  name: clusterName
  identity: {
    type: 'SystemAssigned'
  }
  location: location
  properties: {}
  dependsOn: [
    keyVault
  ]
}

resource keyVaultName_domainAdminSecret 'Microsoft.KeyVault/vaults/secrets@2021-06-01-preview' = {
  parent: keyVault
  name: '${domainAdminSecretName}'
  location: location
  scale: null
  properties: {
    contentType: 'Secret'
    value: domainAdminSecretValue
    attributes: {
      enabled: true
    }
  }
}

resource keyVaultName_localAdminSecret 'Microsoft.KeyVault/vaults/secrets@2021-06-01-preview' = {
  parent: keyVault
  name: '${localAdminSecretName}'
  location: location
  scale: null
  properties: {
    contentType: 'Secret'
    value: localAdminSecretValue
    attributes: {
      enabled: true
    }
  }
}

resource keyVaultName_arbDeploymentSpn 'Microsoft.KeyVault/vaults/secrets@2021-06-01-preview' = {
  parent: keyVault
  name: '${arbDeploymentSpnName}'
  location: location
  scale: null
  properties: {
    contentType: 'Secret'
    value: arbDeploymentSpnValue
    attributes: {
      enabled: true
    }
  }
}

resource keyVaultName_storageWitness 'Microsoft.KeyVault/vaults/secrets@2021-06-01-preview' = {
  parent: keyVault
  name: '${storageWitnessName}'
  location: location
  scale: null
  properties: {
    contentType: 'Secret'
    value: storageWitnessValue
    attributes: {
      enabled: true
    }
  }
}

resource clusterName_default 'microsoft.azurestackhci/clusters/deploymentSettings@[parameters(\'apiVersion\')]' = {
  name: '${clusterName}/default'
  properties: {
    arcNodeResourceIds: arcNodeResourceIds
    deploymentMode: deploymentMode
    deploymentConfiguration: {
      version: '10.0.0.0'
      scaleUnits: [
        {
          deploymentData: {
            securitySettings: {
              hvciProtection: true
              drtmProtection: true
              driftControlEnforced: driftControlEnforced
              credentialGuardEnforced: credentialGuardEnforced
              smbSigningEnforced: smbSigningEnforced
              smbClusterEncryption: smbClusterEncryption
              sideChannelMitigationEnforced: true
              bitlockerBootVolume: bitlockerBootVolume
              bitlockerDataVolumes: bitlockerDataVolumes
              wdacEnforced: wdacEnforced
            }
            observability: {
              streamingDataClient: streamingDataClient
              euLocation: euLocation
              episodicDataUpload: episodicDataUpload
            }
            cluster: {
              name: clusterName
              witnessType: 'Cloud'
              witnessPath: ''
              cloudAccountName: ClusterWitnessStorageAccountName
              azureServiceEndpoint: 'core.windows.net'
            }
            storage: {
              configurationMode: configurationMode
            }
            namingPrefix: namingPrefix
            domainFqdn: domainFqdn
            infrastructureNetwork: [
              {
                subnetMask: subnetMask
                gateway: defaultGateway
                ipPools: [
                  {
                    startingAddress: startingIPAddress
                    endingAddress: endingIPAddress
                  }
                ]
                dnsServers: dnsServers
              }
            ]
            physicalNodes: physicalNodesSettings
            hostNetwork: {
              intents: intentList
              storageNetworks: storageNetworkList
              storageConnectivitySwitchless: storageConnectivitySwitchless
            }
            adouPath: adouPath
            secretsLocation: secretsLocation
            optionalServices: {
              customLocation: customLocation
            }
          }
        }
      ]
    }
  }
  dependsOn: [
    cluster
  ]
}
```

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 3]
```diff
@@ -0,0 +1,1632 @@
...
+
+function Set-NICs {
+    Param (
+        $HCIBoxConfig,
+        [PSCredential]$Credential
+    )
+
+    Invoke-Command -VMName $HCIBoxConfig.MgmtHostConfig.Hostname -Credential $Credential -ScriptBlock {
+        Get-NetAdapter ((Get-NetAdapterAdvancedProperty | Where-Object {$_.DisplayValue -eq "SDN"}).Name) | Rename-NetAdapter -NewName FABRIC
+        # Get-NetAdapter ((Get-NetAdapterAdvancedProperty | Where-Object {$_.DisplayValue -eq "SDN2"}).Name) | Rename-NetAdapter -NewName FABRIC2
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:585]: Can this be removed if not used?

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 4]
```diff
@@ -0,0 +1,1632 @@
...
+    Invoke-Command -VMName $HCIBoxConfig.MgmtHostConfig.Hostname -Credential $localCred -ScriptBlock {
+        $localCred = $using:localCred
+        $domainCred = $using:domainCred
+        $HCIBoxConfig = $using:HCIBoxConfig
+
+        $ErrorActionPreference = "Stop"
+
+        # Disable Fabric2 Network Adapter
+        # Write-Host "Disabling Fabric2 Adapter"
+        # Get-NetAdapter FABRIC2 | Disable-NetAdapter -Confirm:$false | Out-Null
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:656]: Can this be removed if not used?

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 5]
```diff
@@ -0,0 +1,1632 @@
...
+            Restart-Service CertSvc
+            Start-Sleep -Seconds 60
+ 
+            #Issue Certificate Template
+            CMD.exe /c "certutil -SetCATemplates +WebServer"
+        }
+
+        # Write-Host "Configuring DHCP scope on DHCP server."
+        # # Set up DHCP scope for Arc resource bridge
+        # Invoke-Command -VMName $HCIBoxConfig.DCName -Credential $using:domainCred -ArgumentList $HCIBoxConfig -ScriptBlock {
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:918]: Can this region be removed if not used?

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 6]
```diff
@@ -0,0 +1,1632 @@
...
+            # Configure Trusted Hosts
+            Write-Host "Configuring Trusted Hosts on $env:COMPUTERNAME"
+            Set-Item WSMan:\localhost\Client\TrustedHosts * -Confirm:$false -Force
+            
+            # Installing Remote Access
+            Write-Host "Installing Remote Access on $env:COMPUTERNAME" 
+            Install-RemoteAccess -VPNType RoutingOnly | Out-Null
+    
+            # Adding a BGP Router to the VM
+            # Write-Host "Creating BGP Router on $env:COMPUTERNAME"
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:1057]: Can this region be removed if not used?

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 7]
```diff
@@ -0,0 +1,1632 @@
...
+            Write-Host "Installing Hyper-V RSAT Tools on $VMName"
+            Install-WindowsFeature -Name RSAT-Hyper-V-Tools -IncludeAllSubFeature -IncludeManagementTools | Out-Null
+            Write-Host "Installing Active Directory RSAT Tools on $VMName"
+            Install-WindowsFeature -Name  RSAT-ADDS -IncludeAllSubFeature -IncludeManagementTools | Out-Null
+            Write-Host "Installing Failover Clustering RSAT Tools on $VMName"
+            Install-WindowsFeature -Name  RSAT-Clustering-Mgmt, RSAT-Clustering-PowerShell -IncludeAllSubFeature -IncludeManagementTools | Out-Null
+            Write-Host "Installing DNS Server RSAT Tools on $VMName"
+            Install-WindowsFeature -Name RSAT-DNS-Server -IncludeAllSubFeature -IncludeManagementTools | Out-Null
+            Install-RemoteAccess -VPNType RoutingOnly | Out-Null
+            # Install-PackageProvider -Name Nuget -MinimumVersion 2.8.5.201 -Force
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:1208]: Can this be removed if not used?

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 [hunk 8]
```diff
@@ -0,0 +1,1632 @@
...
+Write-Host "[Build cluster - Step 7/10] Build router VM..." -ForegroundColor Green
+New-RouterVM -HCIBoxConfig $HCIBoxConfig -localCred $localCred
+
+# Provision Domain controller VM on AzSMGMT
+Write-Host "[Build cluster - Step 8/10] Building Domain Controller VM..." -ForegroundColor Green
+New-DCVM -HCIBoxConfig $HCIBoxConfig -localCred $localCred -domainCred $domainCred
+
+# Provision Admincenter VM
+# Write-Host "[Build cluster - Step 9/12] Building Windows Admin Center gateway server VM... (skipping step)" -ForegroundColor Green
+#New-AdminCenterVM -HCIBoxConfig $HCIBoxConfig -localCred $localCred -domainCred $domainCred
```
Comment (janegilring) [inline azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:1612]: Can this be removed if not used? Or should the Admin Center VM be installed by this command?

**Review by janegilring — CHANGES_REQUESTED** [review]
Great work 👍

I added some comments which should be reviewed before merging.

File: azure_jumpstart_hcibox/bicep/main.bicep [hunk 9]
```diff
@@ -8,6 +8,9 @@ param spnClientSecret string
 @description('Azure AD tenant id for your service principal')
 param spnTenantId string
 
+@description('Azure AD object id for your Microsoft.AzureStackHCI resource provider')
+param spnProviderId string
```
[the remark this answers is not in the document]
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/bicep/main.bicep:12]: Yes, we do this already with AZD deployment. For a future release we can investigate retrieving it for all deployments.

File: azure_jumpstart_hcibox/artifacts/PowerShell/Configure-VMManagement.ps1 [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+$subId = $env:subscriptionId
+$rg = $env:resourceGroup
+$spnClientId = $env:spnClientId
+$spnSecret = $env:spnClientSecret
+$spnTenantId = $env:spnTenantId
+$location = "eastus"
+$customLocName = $HCIBoxConfig.rbCustomLocationName
+
+# Copy gallery VHDs to hosts
+# Invoke-Command -VMName $HCIBoxConfig.NodeHostConfig[0].Hostname -Credential $adcred -ScriptBlock {
```
[the remark this answers is not in the document]
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/Configure-VMManagement.ps1:31]: yes, in some cases

File: azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1 [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+# Import Configuration Module
+$HCIBoxConfig = Import-PowerShellDataFile -Path $Env:HCIBoxConfigFile
+Start-Transcript -Path "$($HCIBoxConfig.Paths.LogsDir)\Generate-ARM-Template.log"
+
+# Connect to Azure
+Write-Host 'Creating credentials and connecting to Azure'
+$azureAppCred = (New-Object System.Management.Automation.PSCredential $env:spnClientID, (ConvertTo-SecureString -String $env:spnClientSecret -AsPlainText -Force))
+Connect-AzAccount -ServicePrincipal -Subscription $env:subscriptionId -Tenant $env:spnTenantId -Credential $azureAppCred
+
+# Check that extensions are finished installing on HCI nodes
```
[the remark this answers is not in the document]
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1:17]: will evaluate today.

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 (hunk 6 above)
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:1057]: leaving for future consideration

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 (hunk 7 above)
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:1208]: yes

File: azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1 (hunk 1 above)
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1:85]: yes

File: azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1 (hunk 3 above)
Comment (dkirby-ms) [reply azure_jumpstart_hcibox/artifacts/PowerShell/New-HCIBoxCluster.ps1:585]: leaving for now

File: azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1 [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    $arcNodeResourceIds += """" + $machine.id + """"
+    $count = $count + 1
+}
+$arcNodeResourceIds += "]"
+
+# Get storage account key and convert to base 64
+$saKeys = Get-AzStorageAccountKey -ResourceGroupName $env:resourceGroup -Name $env:stagingStorageAccountName
+$storageAccountAccessKey =  [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($saKeys[0].value))
+ 
+# Convert user credentials to base 64
```
Comment (likamrat) [inline azure_jumpstart_hcibox/artifacts/PowerShell/Generate-ARM-Template.ps1:53]:
```suggestion
# Convert user credentials to base64 
```

**Review by likamrat — APPROVED** [review]
LGTM. minor typo

**Review by janegilring — APPROVED** [review]
LGTM

**Review by KidiIT — COMMENTED** [review]
Reviewed all documents and files that were submitted to me and I have been working on them all day and have approved them all for the same amount of money as I have received from the previous owners of the property and the property that was listed on the contract with the seller. Also reviewed the contract with the buyers agreement with them to make it look as good to me to get them out of the deal with me. Thank you very much for your patience and understanding🫶🤓❤️

**Comment by KidiIT** [comment]
All checked out and approved🌚❤️

**Pull request merged**